DBS Issue fix for RTC_498253 RTC_518125 RTC_518127

Gsirisha12345 opened the pull request:
_The PR review is to check for sustainability and correctness.  Sustainability is actually more business critical as correctness is largely tested into the code over time.   Its useful to keep in mind that SW often outlives the HW it was written for and engineers move from job to job so it is critical that code developed for Intel be supportable across many years.  It is up to the submitter and reviewer to look at the code from a perspective of what if we have to debug this 3 years from now after the author is no longer available and defect databases have been lost.  Yes, that happens all the time when we are working with time scales of more than 2 years.  When reviewing your code it is important to look at it from this perspective._

Author Mandatory (to be filled by PR Author/Submitter)
------------------------------------------------------
- Developer who submits the Pull Request for merge is required to mark the checklist below as applicable for the PR changes submitted.  
- Those checklist items which are not marked are considered as not applicable for the PR change.  
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the developers for Inner Source Development Model (ISDM) compliance

### PULL DESCRIPTION
Fix for DBS not removing failed container and failed image, Missing DBS log when docker image is removed, duplicate of DBS remove operation in docker-compose OTA.


### REFERENCES
Reference URL for issue tracking (JIRA/HSD/Github): https://rtc.intel.com/ccm0006001/web/projects/SSEA-%20IOTG%20PED%20Team#action=com.ibm.team.workitem.viewWorkItem&id=498253
https://rtc.intel.com/ccm0006001/web/projects/SSEA-%20IOTG%20PED%20Team#action=com.ibm.team.workitem.viewWorkItem&id=518125
https://rtc.intel.com/ccm0006001/web/projects/SSEA-%20IOTG%20PED%20Team#action=com.ibm.team.workitem.viewWorkItem&id=518127

- [x] **_Added label to the Pull Request following the template: ISDM\_\<Complexity>\*_** \
	Note-1: Depending on complexity of code changes, use the suitable word for complexity: Low/Medium/High \
	Example: PR for Slim boot loader project with medium complexity can have the label as: ISDM_Medium	
- [x] Added label to the Pull Request for easier discoverability and search
- [x] RTC or HSD number will be included in final merge. HSD must always be included if available.
- [x] Changelogs are updated (or N/A if not customer visible)
- [ ] inbm/log_changes.txt and inbm-vision/log_changes.txt are updated for potentially Validation-breaking log changes (or N/A if none)

### CODE MAINTAINABILITY
- [x] **_Commit Message meets guidelines as indicated in the URL\*_**
	- (https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md)
- [x] **_Every commit is a single defect fix and does not mix feature addition or changes\*_**
- [x] Added required new tests relevant to the changes
	- [ ] PR contains URL links to functional tests executed with the new tests 
- [ ] Updated Documentation as relevant to the changes
- [ ] Updated Build steps/commands changes as relevant
- [ ] PR change contains code related to security
- [ ] PR introduces changes that breaks compatibility with other modules (If YES, please provide description)
- [ ] Specific instructions or information for code reviewers (If any):
- [ ] Run 'go fmt' or autopep8 as applicable.
- [ ] New/modified methods and functions should have type annotations on signatures as applicable
- [ ] New/modified methods must have appropriate doc strings (language dependent)

### APPLICATION SPECIFIC
- [ ] Does PR change default config files under /etc? If so, will application still work after an upgrade that leaves /etc alone, like a Mender upgrade?
- [ ] Is cloud UI changed? If so, are cloud definition files updated?



Maintainer Mandatory (to be filled by PR Reviewer/Approving Maintainer)
-----------------------------------------------------------------------
- Maintainer who approves the Pull Request for merge is required to mark the checklist below as appropriate for the PR change reviewed as key proof of attestation indicating reasons for merge. 
- Those checklist items which are not marked are considered as not applicable for the PR change. 
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the maintainers for ISDM compliance.

### QUALITY CHECKS
- [ ] Architectural and Design Fit
- [ ] **_Quality of code (At least one should be checked as applicable)\*_**
	- [ ] Commit Message meets guidelines
	- [ ] PR changes adhere to industry practices and standards
	- [ ] Error and exception code paths implemented correctly
	- [ ] Code reviewed for domain or language specific anti-patterns
	- [ ] Code is adequately commented
	- [ ] Code copyright is correct
	- [ ] Confusing logic is explained in comments
	- [ ] Commit comment can be used to design a new test case for the changes
- [ ] **_Test coverage shows adequate coverage with required CI functional tests pass on all supported platforms\*_**
- [ ] **_Static code scan report shows zero critical issues\*_**
- [ ] Integration tests are passing

### CODE REVIEW IMPACT
- Summary of Defects Detected in Code Review: **\<%P1*xx,P2*xx,P3*xx,P4*xx%>** \
Note P1/P2/P3/P4 denotes severity of defects found (Showstopper/High/Medium/Low) and xx denotes number of defects found

### SECURITY CHECKS
Please check if your PR fulfills the following requirements:

- [ ] Follow best practices when handling primitive data types
- [ ] Configure minimal permissions when opening pipes and ports
- [ ] Check contents within input structures are valid before use
- [ ] All forms of input validated
- [ ] Avoid inter-process race conditions
- [ ] Error and exception handling implemented
- [ ] Defend against Canonical Representation Issues - Any paths utilized?
- [ ] Follow 'secure by default' - Any configuration values added
- [ ] Fail safe - Any failure scenarios?
- [ ] Clean up temporary files - Any temporary files being used?

# _Code must act as a teacher for future developers_

gblewis1 commented on inbm-lib/inbm_lib/trtl.py:
```diff
@@ -403,17 +386,16 @@ def image_remove_by_id(self, image_id: str, force: bool = False) -> Tuple[str, O
+        logger.debug("Trtl.ImageRemoveAll(" + image_id + ")")
         out, err, code = self.runner.run(self._boilerplate(
-            "imageRemoveByID") + " -id=" + image_id + " -f=" + str(force))
+            "ImageRemoveAll") + " -in=" + image_id + " -f=" + str(force))
```
Here we are passing in an image_id to an instanceName parameter. Is this correct? It doesn't look right.

Gsirisha12345 replied:
image_id is not an id, it is a Name of Image. command ImageRemoveAll takes the image name as a parameter, based on that we changed it.
Please refer https://github.com/intel/intel-inb-manageability/blob/develop/inbm/trtl/README.md

gblewis1 replied:
@Gsirisha12345 please have image_remove_by_id remove by id and add a different function image_remove_by_name for removing by name. Otherwise this is confusing for later maintainers.

gblewis1 commented on inbm-lib/tests/unit/inbm_lib/test_trtl.py:
```diff
@@ -158,7 +158,7 @@ def test_trtl_image_remove_by_id(self):
             "abc123", True)
         self.__check_trtl_output("/usr/bin/trtl -type=docker "
-                                 "-cmd=imageRemoveByID -id=abc123 -f=True",
+                                 "-cmd=ImageRemoveAll -in=abc123 -f=True",
```
Same comment here regarding -id vs -in

Gsirisha12345 replied:
image_id is not an id, it is a Name of Image. command ImageRemoveAll takes the image name as a parameter, based on that we changed it.

gblewis1 commented on inbm/diagnostic-agent/diagnostic/event_watcher.py:
```diff
@@ -51,6 +50,7 @@ def run_docker_bench_security(self):  # pragma: no cover
         """Launch Docker Bench Security in separate thread."""
         def run():
             if current_dbs_mode != ConfigDbs.OFF:
+                self.lock.acquire()
```
What happens when the thread is waiting here? Does it block a caller or message sender upstream?

Gsirisha12345 replied:
It will not block anything it will be in queue. Before executing run method, it can acquire the lock. 
After executing run method, it releases the lock. 
Other thread will wait for the lock to be released before they executing the run method.

gblewis1 commented on inbm/dispatcher-agent/dispatcher/remediationmanager/remediation_manager.py:
```diff
@@ -138,21 +130,15 @@ def _remove_container(self, ids: Any) -> None:
                         raise ValueError('Cannot read image ID')
-
-                (out, err, code) = trtl.stop_by_id(str(container_id))
+                logger.debug(container_id)
```
What is the difference between stop_all and stop_all_by_id?

Gsirisha12345 replied:
If any Container/Image has been failed, before going to removal, we should stop the container. 
stop_all Container to be stopped by parsing container Name as a parameter.
stop_by_id stopping container by ID but parsing Name as a parameter instead of Id, so that unable to stop the failed containers.

gblewis1 reviewed (changes requested):
Please address comments. Also provide high level explanation of what the fix is and how it works.

gblewis1 reviewed (changes requested):
Please provide description of test performed to validate this PR and test results. Also provide in the PR a description of root cause of the bug and the approach used to solve it, since this is a complicated issue that has been pending for a long time.

Gsirisha12345 commented:
> Please provide description of test performed to validate this PR and test results. Also provide in the PR a description of root cause of the bug and the approach used to solve it, since this is a complicated issue that has been pending for a long time.

https://rtc.intel.com/ccm0006001/web/projects/SSEA-%20IOTG%20PED%20Team#action=com.ibm.team.workitem.viewWorkItem&id=518127
https://rtc.intel.com/ccm0006001/web/projects/SSEA-%20IOTG%20PED%20Team#action=com.ibm.team.workitem.viewWorkItem&id=518125

Root cause:

From the trtl level, previously they were trying to remove the failed images, by sending image Id to the command (ImageRemoveID), because of that we are not able to remove failed image. Below mentioned code snippet:
out, err, code = self.runner.run(self._boilerplate("imageRemoveByID") + " -id=" + image_id + " -f=" + str(force))
Now we are passing the failed image name to the trtl command (ImageRemoveAll) because of this change we are able to remove the failed image. Below mentioned code snippet:
out, err, code = self.runner.run(self._boilerplate("ImageRemoveAll") + " -in=" + image_id + " -f=" + str(force))
Here, image_id is not an id, it is a Name of Image. cmd (ImageRemoveAll) takes the image name as a parameter, based on that we changed it.

If any Container/Image has been failed, before going to removal, we should stop the container.
stop_by_id stops a container by ID but passing a Name as a parameter instead of an Id, because of that we are unable to stop the failed containers. Mentioned code snippet: (out, err, code) = trtl.stop_by_id(str(container_id))
Now we are passing the stop_all command because of this change we are able to stop the failed containers. Mentioned code snippet: (out, err, code) = trtl.stop_all(str(container_id))

We removed the previous fixed PR's : (RTC 498253) Fix duplicate of DBS remove operation in docker-compose OTA #173 and (RTC 498253) Added a log message if docker image is not present in co… #179

Result: With those changes we are able to remove failed containers and images and logs are not missing.
Fixed RTC-518125 and RTC-518127.

https://rtc.intel.com/ccm0006001/web/projects/SSEA-%20IOTG%20PED%20Team#action=com.ibm.team.workitem.viewWorkItem&id=498253

Root cause:
In addition to the above changes, we added a locking mechanism, in run_docker_bench_security() from the event watcher.
Before executing run method, it can acquire the lock. After executing run method, it releases the lock.
Other thread will wait for the lock to be released before they executing the run method.

Result: We are not getting the duplicate of DBS remove operation in docker-compose OTA. Fixed RTC-498253.

<img width="824" alt="DBS_FIX_TB_RTC518125" src="https://user-images.githubusercontent.com/114566519/208066133-dd8b4db4-9f4c-4883-89e3-58ebd054d29a.png">
<img width="703" alt="DBS_FIX_TB_RTC518127" src="https://user-images.githubusercontent.com/114566519/208066138-b8a6fccb-e549-40c6-8ad9-33df195a4b99.png">
<img width="693" alt="DBS_FIX_TB_RTC498253" src="https://user-images.githubusercontent.com/114566519/208066145-1425cdb0-0970-44cd-996c-0eb8834bdd2a.png">

gblewis1 reviewed (changes requested):
Please have image_remove_by_id remove by id and add a different function image_remove_by_name for removing by name. Otherwise this is confusing for later maintainers.

Gsirisha12345 commented:
> Please provide description of test performed to validate this PR and test results. Also provide in the PR a description of root cause of the bug and the approach used to solve it, since this is a complicated issue that has been pending for a long time.

We found a function (image_remove_all) from the trtl level. cmd (ImageRemoveAll) takes an image name as a parameter to remove failed images, using the (image_remove_all) function instead of adding a new function.
